### ISSUE TITLE
Extension does not detect installed ipykernel, uses wrong pip to try to install

### ISSUE THREAD

**Issue by fonnesbeck** [microsoft/vscode-jupyter#5209]
## Environment data

-   VS Code version: 1.54.3
-   Jupyter Extension version (available under the Extensions sidebar): v2021.3.619093157
-   Python Extension version (available under the Extensions sidebar): v2021.3.658691958
-   OS (Windows | Mac | Linux distro) and version: macOS 11.2.3
-   Python and/or Anaconda version: 3.8.2
-   Type of virtual environment used (N/A | venv | virtualenv | conda | ...): venv
-   Jupyter server running: Local 

## Steps to reproduce:

I am running a notebook in VSCode using a venv that has ipykernel installed, but the extension does not detect it and tries to install it anyway. Unfortunately, it tries to use pip rather than pip3, which results in a failure:

```
❯ /Users/cfonnesbeck/tensorflow_venv/bin/python /Users/cfonnesbeck/.vscode/extensions/ms-python.python-2021.3.658691958/pythonFiles/pyvsc-run-isolated.py /Users/cfonnesbeck/.vscode/extensions/ms-python.python-2021.3.658691958/pythonFiles/shell_exec.py /Users/cfonnesbeck/tensorflow_venv/bin/python /Users/cfonnesbeck/.vscode/extensions/ms-python.python-2021.3.658691958/pythonFiles/pyvsc-run-isolated.py pip install -U ipykernel --user /var/folders/40/_ljkntgn08g0vn2_s9gbxs340000gn/T/tmp-26279FPru7OEEFltb.log
Executing command in shell >> /Users/cfonnesbeck/tensorflow_venv/bin/python /Users/cfonnesbeck/.vscode/extensions/ms-python.python-2021.3.658691958/pythonFiles/pyvsc-run-isolated.py pip install -U ipykernel --user
ERROR: Can not perform a '--user' install. User site-packages are not visible in this virtualenv.
Traceback (most recent call last):
  File "/Users/cfonnesbeck/.vscode/extensions/ms-python.python-2021.3.658691958/pythonFiles/shell_exec.py", line 26, in <module>
    subprocess.check_call(shell_args, stdout=sys.stdout, stderr=sys.stderr)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/Users/cfonnesbeck/tensorflow_venv/bin/python', '/Users/cfonnesbeck/.vscode/extensions/ms-python.python-2021.3.658691958/pythonFiles/pyvsc-run-isolated.py', 'pip', 'install', '-U', 'ipykernel', '--user']' returned non-zero exit status 1.
```

running with pip3 confirms that it is installed in the venv, but the extension still does not detect it. 

Again, I can guarantee that ipykernel is installed in the venv that I am using:

```
❯ ls ~/tensorflow_venv/lib/python3.8/site-packages/ipykernel*
/Users/cfonnesbeck/tensorflow_venv/lib/python3.8/site-packages/ipykernel_launcher.py

/Users/cfonnesbeck/tensorflow_venv/lib/python3.8/site-packages/ipykernel:
__init__.py         displayhook.py      jsonutil.py         resources
__main__.py         embed.py            kernelapp.py        serialize.py
_eventloop_macos.py eventloops.py       kernelbase.py       tests
_version.py         gui                 kernelspec.py       trio_runner.py
codeutil.py         heartbeat.py        log.py              zmqshell.py
comm                inprocess           parentpoller.py
connect.py          iostream.py         pickleutil.py
datapub.py          ipkernel.py         pylab

/Users/cfonnesbeck/tensorflow_venv/lib/python3.8/site-packages/ipykernel-5.5.0.dist-info:
COPYING.md    METADATA      REQUESTED     top_level.txt
INSTALLER     RECORD        WHEEL
```

**Comment by DonJayamanne** [comment]
Thanks for filing this issue.

Please could you enable logging as follows:
* Open settings & go into `Jupyter->Logging`
* Change the value to `debug`
* Reload VS Code,
* Attempt to repro the issue & then copy the output from the `Jupyter` output panel.

After you have the logs from `Jupyter` panel, please try the following:
* Install VS Code Insiders
* Install the Jupyter extension
* Try this once again

Basically when you install the Jupyter extension in VS Code Insiders, you'll get the latest development version of the Jupyter extension & we have fixed a number of issues related to IPYkernel in tehre. 
If the problem persists, we'll need to dig into this further.

**Comment by fonnesbeck** [comment]
Where is the Jupyter output panel? There is a debug console alongside the terminal, but there is nothing written to this console when the problem occurs.

**Comment by DonJayamanne** [comment]
Its here
![Screen Shot 2021-03-19 at 11 22 45](https://user-images.githubusercontent.com/1948812/111826044-73e06800-88a5-11eb-9795-d6c339803733.png)

**Comment by cpoptic** [comment]
Definitely does not fix the issue at all.  Pretty frustrating.  Assured that `ipykernel` is installed in the Conda environment.

**Comment by greazer** [comment]
Likely a dupe of https://github.com/microsoft/vscode-python/issues/15503

**Comment by DonJayamanne** [comment]
@cpoptic

Please could you install VS Code Insiders and let us know if this fixes it
FYI - We believe we have fixed this in our development branch. We publish regular builds from our dev branch to the marketplace. This is only available in VS Code Insiders. Hence testing the dev branch changes/fixes would require you to install VSCode Insiders.

Thanks & do let us know if this fixes it.
If not, please could you enable logging as follows:
* Open settings & go into `Jupyter->Logging`
* Change the value to `debug`
* Reload VS Code,
* Attempt to repro the issue & then copy the output from the `Jupyter` output panel.

**Comment by cpoptic** [comment]
Info 2021-03-29 16:16:55: Execute Cells request 1 0
Info 2021-03-29 16:16:55: Execute Cell 0 file:///d%3A/countyhealthrankings/analysis/cardiovascular_disease_prediction.ipynb in kernelWithMetadata.ts
Info 2021-03-29 16:16:55: Starting Notebook in kernel.ts id = .python3710jvsc74a57bd0d5bf9630585a2cd07048ccbcd2ae7d2b8b80044640aa0d13db29545513b9c097.C:/ProgramData/Anaconda3/envs/countyhealth\python.exe.C:\ProgramData\Anaconda3\envs\countyhealth\python.exe.Python 3.7.10 64-bit ('countyhealth': conda)
Info 2021-03-29 16:16:55: Cell Index:0, state:1, exec: undefined. User queued cell for execution
Info 2021-03-29 16:16:55: Cell Index:0, state:1, exec: undefined. Update cell state as it was enqueued
Info 2021-03-29 16:16:55: Creating raw notebook for file:///d%3A/countyhealthrankings/analysis/cardiovascular_disease_prediction.ipynb
Info 2021-03-29 16:16:55: Getting preferred kernel for file:///d%3A/countyhealthrankings/analysis/cardiovascular_disease_prediction.ipynb
Info 2021-03-29 16:16:55: Computing working directory file:///d%3A/countyhealthrankings/analysis/cardiovascular_disease_prediction.ipynb
Info 2021-03-29 16:16:55: Connecting to raw session for file:///d%3A/countyhealthrankings/analysis/cardiovascular_disease_prediction.ipynb with connection {"kind":"startUsingPythonInterpreter","kernelSpec":{"specFile":"C:\\ProgramData\\Anaconda3\\envs\\countyhealth\\share\\jupyter\\kernels\\python3\\kernel.json","interpreterPath":"C:\\ProgramData\\Anaconda3\\envs\\countyhealth\\python.exe","name":"python3710jvsc74a57bd0d5bf9630585a2cd07048ccbcd2ae7d2b8b80044640aa0d13db29545513b9c097","argv":["C:/ProgramData/Anaconda3/envs/countyhealth\\python.exe","-m","ipykernel_launcher","-f","{connection_file}"],"language":"python","path":"C:/ProgramData/Anaconda3/envs/countyhealth\\python.exe","display_name":"Python 3.7.10 64-bit ('countyhealth': conda)"},"interpreter":{"architecture":3,"path":"C:\\ProgramData\\Anaconda3\\envs\\countyhealth\\python.exe","version":{"options":{"loose":false,"includePrerelease":false},"loose":false,"raw":"3.7.10-final","major":3,"minor":7,"patch":10,"prerelease":["final"],"build":[],"version":"3.7.10-final"},"sysVersion":"3.7.10 (default, Feb 26 2021, 13:06:18) [MSC v.1916 64 bit (AMD64)]","sysPrefix":"C:\\ProgramData\\Anaconda3\\envs\\countyhealth","fileHash":"0ea6f48487af49960138a0a437cc02502a446960b87640f76bb458eae6545104e940b7d3d067bd5ec0eb226ab7d1affdc3ebf7a3a4cd008eb15c97de1aa7eee6","companyDisplayName":"Anaconda, Inc.","envType":"Conda","envPath":"C:\\ProgramData\\Anaconda3\\envs\\countyhealth","envName":"countyhealth","cachedEntry":true,"displayName":"Python 3.7.10 64-bit ('countyhealth': conda)"},"id":".python3710jvsc74a57bd0d5bf9630585a2cd07048ccbcd2ae7d2b8b80044640aa0d13db29545513b9c097.C:/ProgramData/Anaconda3/envs/countyhealth\\python.exe.C:\\ProgramData\\Anaconda3\\envs\\countyhealth\\python.exe.Python 3.7.10 64-bit ('countyhealth': conda)"}
Info 2021-03-29 16:16:55: Starting raw kernel Python 3.7.10 64-bit ('countyhealth': conda)
Info 2021-03-29 16:16:55: installMissingDependencies C:\ProgramData\Anaconda3\envs\countyhealth\python.exe
Info 2021-03-29 16:16:55: Cached data exists getEnvironmentVariables, <No Resource>
Info 2021-03-29 16:16:55: Process Execution: > C:\ProgramData\Anaconda3\envs\countyhealth\python.exe c:\Users\Chris\.vscode-insiders\extensions\ms-toolsai.jupyter-2021.5.697286501\pythonFiles\pyvsc-run-isolated.py -c "import ipykernel"
> C:\ProgramData\Anaconda3\envs\countyhealth\python.exe c:\Users\Chris\.vscode-insiders\extensions\ms-toolsai.jupyter-2021.5.697286501\pythonFiles\pyvsc-run-isolated.py -c "import ipykernel"
Info 2021-03-29 16:16:55: Process Execution: > C:\ProgramData\Anaconda3\envs\countyhealth\python.exe c:\Users\Chris\.vscode-insiders\extensions\ms-toolsai.jupyter-2021.5.697286501\pythonFiles\pyvsc-run-isolated.py -c "import ipykernel"
> C:\ProgramData\Anaconda3\envs\countyhealth\python.exe c:\Users\Chris\.vscode-insiders\extensions\ms-toolsai.jupyter-2021.5.697286501\pythonFiles\pyvsc-run-isolated.py -c "import ipykernel"

**Comment by DonJayamanne** [comment]
@cpoptic 

I'm sorry you're having issues with this extension.
We're trying our best to ensure this works well. Unfortunately this is challenging. We've had a lot of issues with Conda & been constantly working around conda, primarily because conda doesn't work very well when not using termnials, partly explains why users MUST run `conda init` in their shells (each shell).

Please could you run this in your terminal & let me know what you get:
`C:\ProgramData\Anaconda3\envs\countyhealth\python.exe c:\Users\Chris.vscode-insiders\extensions\ms-toolsai.jupyter-2021.5.697286501\pythonFiles\pyvsc-run-isolated.py -c "import ipykernel"`

**Comment by greazer** [comment]
Closing due to no response in over a week.
